# EDA on CHAPA Chapter 40B Application Data from Oct 2023 to May 2025
Author: Gia Bao (Belinda) Ngo

**Dataset:** TOUCHABLE CHAPA Chapter 40B Application Data 2021-2023 & 10|2023-05|2025 (https://docs.google.com/spreadsheets/d/1xcW0qveqOPD-1JvJX1jaSJjPyCVVMGbY/edit?gid=690813037#gid=690813037)


**Sheet used:** `23-25_data_updated`

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Load and choose dataset
### Load Excel file and explore 23-25 sheets

In [2]:
file_path = "../data/TOUCHABLE CHAPA Chapter 40B Application Data 2021-2023 & 10_2023-05_2025.xlsx"

# list all sheet names in the Excel file
for sheet in pd.ExcelFile(file_path).sheet_names:
    print(f"{sheet}")

2023-2025
Saniya 2023-2025
Clean Saniya 2023-2025
Copy Clean Saniya 23-25
23-25_data_updated
NEW 2023-2025
2021-2023
CLEAN Kayla 2021-2023
Copy of CLEAN Kayla 2021-2023
Geocoded Kayla 2021-2023
Kayla Working 2021-2023
Kayla 2021-2023 Working 2.0
CLEAN w stats Kayla 2021-2023


In [33]:
# # look at shape of all 23-25 sheets
# sheet_list = ["2023-2025",
#               "Saniya 2023-2025",
#               "Clean Saniya 2023-2025",
#               "Copy Clean Saniya 23-25",
#               "23-25_data_updated",
#               "NEW 2023-2025"]

# for sheet in sheet_list:
#     print(f"Sheet: {sheet}")
#     df = pd.read_excel(file_path, sheet_name=sheet)
#     print(df.shape, "\n")

# # look at info of 23-25 sheets with more columns
# sheet_list = ["Saniya 2023-2025",
#               "Clean Saniya 2023-2025",
#               "Copy Clean Saniya 23-25",
#               "23-25_data_updated"]

# for sheet in sheet_list:
#     print(f"Sheet: {sheet}")
#     df = pd.read_excel(file_path, sheet_name=sheet)
#     print(df.info(), "\n")

# # compare column names between sheets
# sheet_list = ["Saniya 2023-2025",
#               "Clean Saniya 2023-2025",
#               "Copy Clean Saniya 23-25",
#               "23-25_data_updated"]

# for sheet in sheet_list:
#     print(f"Sheet: {sheet}")
#     df = pd.read_excel(file_path, sheet_name=sheet)
#     print(df.columns.to_list(), "\n")

| Sheet name | Shape | Notes |
|----------|----------|----------|
| 2023-2025 | (1360, 14) | most likely raw data |
| Saniya 2023-2025 | (1360, 31) | later columns contain stats, still in cleaning process |
| Clean Saniya 2023-2025 | (1360, 29) | duplicate columns, still in cleaning process |
| Copy Clean Saniya 23-25 | (1360, 20) | has "NA" values, race data from above sheet standardized here |
| 23-25_data_updated | (1359, 29) | NA values are blanks here |
| NEW 2023-2025 | (1360, 14) | most likely raw data |

### Load chosen dataset: `23-25_data_updated`
`shape`: (1359, 29)

In [96]:
df = pd.read_excel(file_path, sheet_name="23-25_data_updated")
df.head()

,ID Number,Submission Date,Submission Time,Application Property,Application City/Town,Application St Address,Application Unit,Source New,Age,Race Combination Normalized,...,FTHB Class?,Unique ID,Street Address,City,State,ZIP,Match Status,Match Type,Matched Address,MSA
0,101,2024-05-28,13:31:01,"14 Caspian Way, Unit 23 ~ Leominster",Leominster,14 Caspian Way,unit 23,MyMassHome,65.0,White,...,No,497,14 Caspian Way,Leominster,MA,NaN,Match,Non_Exact,"14 CASPIAN ST, LEOMINSTER, MA, 01453","Worcester, MA"
1,101,2024-07-21,17:33:34,1 Millbrook Lane Unit 213 ~ Wakefield,Wakefield,1 Millbrook Lane Unit 213,NaN,MyMassHome,65.0,White,...,No,696,1 Millbrook Lane Unit 213,Wakefield,MA,NaN,Match,Exact,"1 MILLBROOK LN, WAKEFIELD, MA, 01880","Boston-Cambridge-Newton, MA-NH"
2,1595,2024-02-05,13:37:00,"11 Indian Woods Way, Unit 11 ~ Canton",Canton,11 Indian Woods Way,unit 11,City of Boston Metrolist,56.0,Black or African American,...,Yes,175,11 Indian Woods Way,Canton,MA,NaN,Match,Exact,"11 INDIAN WOODS WAY, CANTON, MA, 02021","Boston-Cambridge-Newton, MA-NH"
3,2159,2024-07-22,23:33:59,24 Haywood Circle ~ Mendon,Mendon,24 Haywood Circle,NaN,Zillow/Trulia/Other Website,30.0,Black or African American,...,Yes,706,24 Haywood Circle,Mendon,MA,NaN,Match,Exact,"24 HAYWOOD CIR, MENDON, MA, 01756","Worcester, MA"
4,2685,2024-08-28,10:50:33,"6 Woodman Way, unit 204A ~ Newburyport",Newburyport,6 Woodman Way,unit 204a,Real Estate Agent,58.0,White,...,No,888,6 Woodman Way,Newburyport,MA,NaN,Match,Exact,"6 WOODMAN WAY, NEWBURYPORT, MA, 01950","Boston-Cambridge-Newton, MA-NH"


## Explore dataset

In [102]:
# standardize column names
df.columns = df.columns.str.strip().str.lower().str.replace(" ", "_").str.replace("/", "_").str.replace("?", "")
# df.columns.to_list()

In [101]:
# check dtypes and nulls
# df.info(), df.isnull().sum().sort_values(ascending=False)

Data types are as expected.

### Explore columns

In [142]:
# id_number
print(f"total count: {len(df["id_number"])}, unique value count: {df["id_number"].nunique()}")

total count: 1359, unique value count: 964


In [143]:
# application_property
df[["application_property", "age"]].iloc[0], df[["application_property", "age"]].iloc[7]

(application_property    14 Caspian Way, Unit 23 ~ Leominster
 age                                                     65.0
 Name: 0, dtype: object,
 application_property    63 Central St., #107 ~ North Reading (55+)
 age                                                           64.0
 Name: 7, dtype: object)

In [147]:
# application_unit
df["application_unit"].value_counts().head(10), df["application_unit"].isnull().sum()

(application_unit
 unit 88       84
 unit 4-302    65
 unit 23       60
 unit 17       51
 unit 11       50
 unit 13       31
 unit 109      25
 unit #4       20
 #a1           19
 unit 110      18
 Name: count, dtype: int64,
 np.int64(765))

In [146]:
# source_new
df["source_new"].value_counts(), df["source_new"].isnull().sum()

(source_new
 MyMassHome                     504
 Zillow/Trulia/Other Website    361
 Real Estate Agent              338
 Friend/Word of Mouth            96
 City of Boston Metrolist        21
 Social Media                    15
 Name: count, dtype: int64,
 np.int64(24))

In [153]:
# age
df["age"].describe(), df["age"].isnull().sum(), df["age"].value_counts().sort_index()

(count    1358.000000
 mean       41.023932
 std        13.610772
 min         0.000000
 25%        31.000000
 50%        38.000000
 75%        49.000000
 max        86.000000
 Name: age, dtype: float64,
 np.int64(1),
 age
 0.0     1
 4.0     1
 19.0    4
 20.0    2
 21.0    7
        ..
 79.0    2
 82.0    1
 83.0    1
 85.0    1
 86.0    1
 Name: count, Length: 68, dtype: int64)

In [160]:
# race_combination_normalized
df["race_combination_normalized"].nunique(), df["race_combination_normalized"].value_counts().head(10)

(26,
 race_combination_normalized
 White                                            666
 Black or African American                        187
 Hispanic or Latino                               149
 Asian                                            130
 Choose not to answer                              51
 Asian, Pacific Islander or Native Hawaiian        44
 Middle Eastern or North African                   34
 Black or African American, Hispanic or Latino     26
 Hispanic or Latino, White                         23
 Black or African American, White                  11
 Name: count, dtype: int64)

In [161]:
# census_races
df["census_races"].nunique(), df["census_races"].value_counts().head(10)

(18,
 census_races
 White alone                                                730
 Missing                                                    203
 Black or African American alone                            187
 Asian alone                                                130
 Asian; Native Hawaiian or Other Pacific Islander            44
 Black or African American                                   26
 Black or African American; White                            14
 Asian; White                                                 8
 Asian; Native Hawaiian or Other Pacific Islander; White      4
 American Indian and Alaska Native alone                      4
 Name: count, dtype: int64)

In [164]:
# census_races
condition = df["census_races"] == "Missing"
columns = ["race_combination_normalized", "census_races"]
df.loc[condition, columns]["race_combination_normalized"].value_counts()

race_combination_normalized
Hispanic or Latino      149
Choose not to answer     51
Missing                   3
Name: count, dtype: int64

In [165]:
# census_races
condition = df["race_combination_normalized"] == "Black or African American, Hispanic or Latino, White"
columns = ["race_combination_normalized", "census_races"]
df.loc[condition, columns]

,race_combination_normalized,census_races
1016,"Black or African American, Hispanic or Latino,...",Black or African American; White
1017,"Black or African American, Hispanic or Latino,...",Black or African American; White
1018,"Black or African American, Hispanic or Latino,...",Black or African American; White


In [166]:
# disability
df["disability"].value_counts(), df["disability"].isnull().sum()

(disability
 No                      1233
 Yes                      118
 Choose not to answer       5
 no                         1
 Name: count, dtype: int64,
 np.int64(2))

In [169]:
# current_residence_town_city
df["current_residence_town_city"].nunique(), df["current_residence_town_city"].isnull().sum()

(251, np.int64(2))

In [172]:
# current_residence_state
df["current_residence_state"].value_counts(), df["current_residence_state"].isnull().sum()

(current_residence_state
 MA               1328
 NH                 12
 RI                  5
 Ma                  5
 AR                  2
 FL                  1
 Massachusetts       1
 SC                  1
 ma                  1
 CT                  1
 ME                  1
 Name: count, dtype: int64,
 np.int64(1))

In [173]:
# hh_size
df["hh_size"].value_counts(), df["hh_size"].isnull().sum()

(hh_size
 1     562
 2     345
 3     218
 4     149
 5      55
 6      18
 0       7
 7       3
 20      1
 31      1
 Name: count, dtype: int64,
 np.int64(0))

In [175]:
# dependents
df["dependents"].value_counts(), df["dependents"].isnull().sum()

(dependents
 0    820
 1    303
 2    174
 3     50
 4     10
 6      1
 7      1
 Name: count, dtype: int64,
 np.int64(0))

In [178]:
# hh_type_standardized
df["hh_type_standardized"].value_counts(), df["hh_type_standardized"].isnull().sum()

(hh_type_standardized
 Single person                                      630
  Married/partners/couple, no dependents            412
 Single parent                                      275
  More than one related adults, with dependents      17
 Other                                               13
  More than one related adults, no dependents        10
  Married/partners/couple, with dependents            2
 Name: count, dtype: int64,
 np.int64(0))

In [ ]:
# hh_type_standardized
condition = df["hh_type_standardized"] == "Other"
columns = ["id_number", "age", "hh_size", "dependents", "hh_type_standardized", "hh_income", "hh_assets"]
df.loc[condition, columns]

,id_number,age,hh_size,dependents,hh_type_standardized,hh_income,hh_assets
40,33704,21.0,2,0,Other,46760.00,42000.0
46,40209,58.0,1,0,Other,875.20,318018.0
114,100881,27.0,2,0,Other,64000.00,69000.0
614,467102,43.0,2,0,Other,40191.12,61100.0
889,658386,23.0,2,0,Other,61000.00,35300.0
890,658386,23.0,2,0,Other,62000.00,34000.0
891,658386,23.0,2,0,Other,60000.00,36000.0
892,658386,23.0,2,0,Other,60000.00,32000.0
893,658386,23.0,2,0,Other,60000.00,32000.0
922,677756,66.0,2,0,Other,28000.00,250000.0


In [187]:
# hh_income
df["hh_income"].isnull().sum(), df["hh_income"].value_counts().sort_index().head(10), df["hh_income"].value_counts().sort_index().tail(10)

(np.int64(5),
 hh_income
 70.00      1
 78.00      1
 96.50      1
 100.00     1
 101.00     2
 150.00     1
 875.20     1
 2115.39    1
 2300.00    2
 3000.00    2
 Name: count, dtype: int64,
 hh_income
 138000.00      1
 147836.31      1
 150000.00      1
 187000.00      2
 213556.00      1
 250000.00      1
 493323.00      1
 607700.00      1
 890000.00      1
 70200405.00    1
 Name: count, dtype: int64)

In [186]:
# hh_income
condition = df["hh_income"] > 150000 #< 12000
columns = ["id_number", "age", "hh_size", "dependents", "hh_income", "hh_assets"]
df.loc[condition, columns]

,id_number,age,hh_size,dependents,hh_income,hh_assets
736,550104,34.0,1,0,607700.0,61362.21
886,653600,66.0,1,0,213556.0,265081.00
903,660769,32.0,1,0,890000.0,0.00
991,726420,54.0,1,0,493323.0,30654.00
1027,753912,28.0,5,1,250000.0,10700.00
1139,841464,63.0,1,0,70200405.0,40500.00
1344,990761,36.0,4,2,187000.0,20000.00
1345,990761,36.0,4,2,187000.0,94000.00


In [189]:
# hh_assets
df["hh_assets"].isnull().sum(), df["hh_assets"].describe()

(np.int64(1),
 count      1358.000000
 mean      43125.715390
 std       56536.667031
 min           0.000000
 25%       13000.000000
 50%       30223.000000
 75%       56971.895000
 max      863967.300000
 Name: hh_assets, dtype: float64)

In [190]:
# fthb_class
df["fthb_class"].value_counts(), df["fthb_class"].isnull().sum()

(fthb_class
 No     746
 Yes    610
 Name: count, dtype: int64,
 np.int64(3))

In [192]:
# unique_id
len(df), df["unique_id"].nunique(), df["unique_id"].isnull().sum()

(1359, 1359, np.int64(0))

In [194]:
# street_address
df["street_address"].nunique(), df["street_address"].isnull().sum(), df["street_address"].value_counts().head(10)

(59,
 np.int64(0),
 street_address
 74 Riley Road            84
 501 Commerce Drive       75
 24 Haywood Circle        69
 27 Meeting House Lane    69
 28 Longview Circle       63
 14 Caspian Way           60
 14 Autumn Lane           59
 15 Jasmine Drive         51
 11 Indian Woods Way      50
 8 Kayak Trail            47
 Name: count, dtype: int64)

In [197]:
# city
df["city"].nunique(), df["city"].isnull().sum(), df["city"].value_counts().head(10)

(45,
 np.int64(0),
 city
 Tyngsborough     84
 North Andover    83
 Ayer             78
 Braintree        75
 Mendon           69
 Southborough     69
 West Boylston    68
 Leominster       64
 Northborough     59
 Canton           50
 Name: count, dtype: int64)

In [198]:
# state
df["state"].value_counts(), df["state"].isnull().sum()

(state
 MA    1359
 Name: count, dtype: int64,
 np.int64(0))

In [200]:
# zip
df["zip"].isnull().sum()

np.int64(1359)

In [201]:
# match_status
df["match_status"].value_counts(), df["match_status"].isnull().sum()

(match_status
 Match       1209
 No_Match     150
 Name: count, dtype: int64,
 np.int64(0))

In [202]:
# match_type
df["match_type"].value_counts(), df["matched_address"].isnull().sum()

(match_type
 Exact        1020
 Non_Exact     189
 Name: count, dtype: int64,
 np.int64(150))

In [203]:
# matched_address
df["matched_address"].isnull().sum(), df["matched_address"].nunique(), df["matched_address"].value_counts().head(10)

(np.int64(150),
 53,
 matched_address
 501 COMMERCE DR, BRAINTREE, MA, 02184           75
 27 MEETING HOUSE LN, SOUTHBOROUGH, MA, 01772    69
 24 HAYWOOD CIR, MENDON, MA, 01756               69
 28 LONGVIEW CIR, AYER, MA, 01432                63
 14 CASPIAN ST, LEOMINSTER, MA, 01453            60
 14 AUTUMN LN, NORTHBOROUGH, MA, 01532           59
 15 JASMINE DR, WEST BOYLSTON, MA, 01583         51
 11 INDIAN WOODS WAY, CANTON, MA, 02021          50
 8 KAYAK TRL, NORTON, MA, 02766                  47
 42 PINE ST, NEWTON, MA, 02465                   46
 Name: count, dtype: int64)

In [205]:
# msa
df["msa"].value_counts(), df["msa"].isnull().sum()

(msa
 Boston-Cambridge-Newton, MA-NH      640
 Worcester, MA                       457
 Boston-Cambridge-Newton, MA–NH      144
 Providence-Warwick, RI-MA            63
 Amherst Town-Northampton, MA         20
 N/A: Dukes County Vineyard Haven     18
 Barnstable Town, MA                  17
 Name: count, dtype: int64,
 np.int64(0))

In [206]:
# msa
condition = df["msa"] == "N/A: Dukes County Vineyard Haven"
columns = ["matched_address", "msa"]
df.loc[condition, columns]

,matched_address,msa
6,"1 MAGNOLIA WAY, EDGARTOWN, MA, 02539",N/A: Dukes County Vineyard Haven
64,"1 MAGNOLIA WAY, EDGARTOWN, MA, 02539",N/A: Dukes County Vineyard Haven
68,"1 MAGNOLIA WAY, EDGARTOWN, MA, 02539",N/A: Dukes County Vineyard Haven
73,"1 MAGNOLIA WAY, EDGARTOWN, MA, 02539",N/A: Dukes County Vineyard Haven
87,"1 MAGNOLIA WAY, EDGARTOWN, MA, 02539",N/A: Dukes County Vineyard Haven
94,"1 MAGNOLIA WAY, EDGARTOWN, MA, 02539",N/A: Dukes County Vineyard Haven
517,"1 MAGNOLIA WAY, EDGARTOWN, MA, 02539",N/A: Dukes County Vineyard Haven
520,"1 MAGNOLIA WAY, EDGARTOWN, MA, 02539",N/A: Dukes County Vineyard Haven
534,"1 MAGNOLIA WAY, EDGARTOWN, MA, 02539",N/A: Dukes County Vineyard Haven
647,"1 MAGNOLIA WAY, EDGARTOWN, MA, 02539",N/A: Dukes County Vineyard Haven


### Notes on all Columns

| Column | Null count | Notes |
|--------|-------------|-------|
| id_number | 0 | total count: 1359, unique value count: 964 |
| submission_date | 0 | |
| submission_time | 0 | |
| application_property | 0 | mapped into next 3 columns; "(55+)", "(age restricted)"; `age` can be >55 but not applying to an age restricted property |
| application_city_town | 0 | |
| application_st_address | 0 | |
| application_unit | 765 | some properties don't have units; "unit", "#" |
| source_new | 24 | |
| age | 1 | 1 "0", 1 "4", else 19 and up |
| race_combination_normalized | 0 | change "Missing" (3) to NA, consider dropping "Other" (1) & changing to list dtype, how to handle "Choose not to answer"? |
| census_races | 0 | all `race_combination_normalized` "Hispanic or Latino", "Choose not to answer", and "Missing" in `census_races` "Missing"; col disregards "Hispanic or Latino" as a race |
| disability | 2 | "no", "No" |
| current_residence_town_city | 2 | shares 1 null row with `age` |
| current_residence_state | 1 | "MA", "Massachusetts", "ma" |
| hh_size | 0 | "20", "31" |
| dependents | 0 | |
| hh_type_standardized | 0 | "Other" |
| hh_income | 5 | outlier values; is there an income requirement? |
| hh_assets | 1 | |
| fthb_class | 3 | shares 1 null row with `age` |
| unique_id | 0 | |
| street_address | 0 | 59 unique values |
| city | 0 | 45 unique values |
| state | 0 | all "MA" |
| zip | 1359 | blank column |
| match_status | 0 | geocoding result comparing property address to `matched_address`, "Match" or "No_Match" |
| match_type | 150 | "Exact", "Non_Exact", NA |
| matched_address | 150 | 53 unique values |
| msa | 0 | |


## Manipulate dataset


ask client about possible `hh_income` outliers

* ask how to handle:
    * `age`: "0", "4", what is age requirement to apply?
    * `race_combination_normalized`: drop "Missing" (3), "Other" (1)?
    * use raw data from `race_combination_normalized` or still adhere to `census_races`?
        * disregards "Hispanic or Latino"
    * "Choose not to answer" in `race_combination_normalized`, `disability`
    * `hh_size`: "20", "31"
    * `hh_type_standardized`: "Other" (13)
    * does CHAPA only look at `hh_income` or `hh_income` + `hh_assets`?
    * limited nulls: `age`, `disability`, `current_residence_town_city`, `current_residence_state`, `hh_income`, `hh_assets`, `fthb_class`

* clean:
    * `application_unit`: filters out "unit" or "#"
    * `race_combination_normalized`: consider changing to list dtype
    * `disability`: standardize "no" to "No"
    * `current_residence_state`: standardize "Massachusetts", "ma", "Ma" to "MA"

* create:
    * `age_restricted`: 1 if indicated as "(55+)" or "(age restricted)" in `application_property`, 0 otherwise
    * `zip`: pull from `matched_address`



In [212]:
# standardize columns
# disability
df["disability"] = df["disability"].replace({"no": "No"})
# current_residence_state
for x in ["Ma", "ma", "Massachusetts"]:
    df["current_residence_state"] = df["current_residence_state"].replace({x: "MA"})
# application_unit
df["application_unit"] = (
    df["application_unit"]
    .str.replace(r'(?i)unit', '', regex=True)   #case-insensitive
    .str.replace('#', '', regex=False)
    .str.strip()
)

Demographics:
* race
* household type
* age
* marketing source
* fthb

## Visualize

ideas
* `id_number`: explore re-applications, note how long between applications
